consensus/polybft/validators_snapshot.go

```diff
@@ -214,29 +224,98 @@ func (v *validatorsSnapshotCache) storeSnapshot(block uint64, snapshot AccountSe
 // Cleanup cleans the validators cache in memory and db
 func (v *validatorsSnapshotCache) cleanup() error {
 	if len(v.snapshots) >= validatorSnapshotLimit {
-		latestBlock := uint64(0)
+		latestEpoch := uint64(0)
 
 		for e := range v.snapshots {
-			if e > latestBlock {
-				latestBlock = e
+			if e > latestEpoch {
+				latestEpoch = e
 			}
 		}
 
-		block := latestBlock
-		cache := make(map[uint64]AccountSet, numberOfSnapshotsToLeaveInMemory)
+		startEpoch := latestEpoch
+		cache := make(map[uint64]*validatorSnapshot, numberOfSnapshotsToLeaveInMemory)
 
 		for i := 0; i < numberOfSnapshotsToLeaveInMemory; i++ {
-			if snapshot, exists := v.snapshots[block]; exists {
-				cache[block] = snapshot
+			if snapshot, exists := v.snapshots[startEpoch]; exists {
+				cache[startEpoch] = snapshot
 			}
 
-			block--
+			startEpoch--
 		}
 
 		v.snapshots = cache
 
-		return v.state.cleanValidatorSnapshotsFromDB(latestBlock)
+		return v.state.cleanValidatorSnapshotsFromDB(latestEpoch)
 	}
 
 	return nil
 }
+
+// getLastCachedSnapshot gets the latest snapshot cached
+// If it doesn't have snapshot cached for desired epoch, it will return the latest one it has
+func (v *validatorsSnapshotCache) getLastCachedSnapshot(currentEpoch uint64) (*validatorSnapshot, error) {
+	cachedSnapshot := v.snapshots[currentEpoch]
+	if cachedSnapshot != nil {
+		return cachedSnapshot, nil
+	}
+
+	// if we do not have a snapshot in memory for given epoch, we will get the latest one we have
+	for ; currentEpoch >= 0; currentEpoch-- {
+		cachedSnapshot = v.snapshots[currentEpoch]
+		if cachedSnapshot != nil {
+			v.logger.Trace("Found snapshot in memory cache", "Epoch", currentEpoch)
+
+			break
+		}
+
+		if currentEpoch == 0 { // prevent uint64 underflow
+			break
+		}
+	}
+
+	dbSnapshot, err := v.state.getLastSnapshot()
+	if err != nil {
+		return nil, err
+	}
+
+	if dbSnapshot != nil {
+		// if we do not have any snapshot in memory, or db snapshot is newer than the one in memory
+		// return the one from db
+		if cachedSnapshot == nil || dbSnapshot.Epoch > cachedSnapshot.Epoch {
+			cachedSnapshot = dbSnapshot
```

Comment:
Don't we need to store the snapshot in the cache at this point?

Reply:
Good point. Fix: https://github.com/0xPolygon/polygon-edge/pull/997/commits/43c3ecea6729a138f43da4378fab505569002594